Allow empty splits on panes

iff opened the pull request:
Draft as a base for continuing the discussion in #8008 : adds a `SplitOperation` enum to support bindings like `["pane::SplitLeft", {"operation": "Clear"}]`

To be discussed @MrSubidubi and others:

- Naming: Generally not happy with names yet and specifically `Empty` is unclear, e.g., what does this mean for terminal panes? Added placeholder code to split without cloning, but unsure what users would expect in this case.
- ~~I removed `SplitAndMoveXyz` actions but I guess we should keep them for backwards compatibility?~~
- May have missed details in the move implementation. Will check the code again for opportunities to refactor more code after we agree on the approach.
- ~~Tests should go to `crates/collab/src/tests/integration_tests.rs`?~~

Closes #8008

Release Notes:

- Add `pane::Split` mode (`{ClonePane,EmptyPane,MovePane}`) to allow creating an empty buffer.

iff committed:
prototype new split clone/empty/move behavior

- adding an enum for SplitBehavior (naming not final) to support bindings like `["pane::SplitLeft", {"behavior": "Empty"}]`
- subsuming SplitAndMoveDirection actions but probably have to keep them (?)
- empty behavior for terminal panes: unclear what this means. Currently split but does not clone.

iff commented:
Yup. Just waiting for feedback.

zelenenka commented:
@iff got it, thanks. i'll ping people internally!

iff commented:
@MrSubidubi no worries at all and completely understandable. Thanks for the feedback - I will work on integrating the feedback and finishing the implementation this week.

iff committed:
Merge remote-tracking branch 'upstream/main' into split-behavior

iff committed:
cargo fmt

iff committed:
fix merge and re-add split and move

iff committed:
pane: sketching some tests

iff committed:
pane: add test helper

iff committed:
pane: add a more generic test helper

iff committed:
rename behavior to operation

iff committed:
rename empty split operation to clear

iff committed:
cleanup

MrSubidubi commented on crates/workspace/src/pane.rs:
```diff
@@ -6981,4 +7135,151 @@ mod tests {
     }
+
+    // Assert the item label, with the active item label expected active index
+    // TODO dont like the duplication
```
At the end of the day, you added tests here, so I don't really mind that. Most people don't do it, you did it properly, would be more than fine with me.

MrSubidubi reviewed (commented):
Two days delayed. sorry. Some thoughts, please ping me if you feel I am missing something.

iff committed:
terminal_panel: matches instead of match/unreachable

Co-authored-by: Finn Evers <finn.evers@outlook.de>

iff committed:
terminal_pane: convention and naming

iff committed:
pane: make SplitOperation Copy

iff commented on crates/workspace/src/pane.rs:
```diff
@@ -196,6 +196,102 @@ pub struct DeploySearch {
+    pub operation: SplitOperation,
+}
+
+// TODO not so sure if we need/want those?
```
Yeah it is used only four times and I agree that does not justify the added code. I will give it another shot to find alternative implementations that feel more ergonomic :).

iff committed:
pane: remove split operations impls

iff committed:
Merge branch 'main' into split-behavior

iff committed:
pane: fix tests

iff committed:
copy SplitOperation

iff committed:
pane: macro for split structs

iff committed:
Merge remote-tracking branch 'upstream/main' into split-behavior

iff committed:
pane: doc strings for SplitOperation and macro

iff committed:
doc needs to be literal

MrSubidubi reviewed (approved):
Very much like the solution you came up with, thank you very much!

One last question that is not necessarily blocking and I think we are good to merge

iff commented on crates/workspace/src/pane.rs:
```diff
@@ -196,6 +196,102 @@ pub struct DeploySearch {
+// TODO naming
+#[derive(Clone, PartialEq, Debug, Deserialize, JsonSchema, Default)]
+#[serde(deny_unknown_fields)]
+pub enum SplitOperation {
```
Sorry I missed this earlier, but I actually prefer your suggested naming. I would make all variants end with `Pane` (e.g., `ClonePane`). Or is `Item` a term you use in connection with panes? `EmptyPane` still feels a bit odd for a terminal pane, since it’s never truly empty (there’s always a prompt), but it’s probably good enough.

iff committed:
rename SplitOperation -> SplitMode

iff committed:
rename split struct field to mode

iff committed:
Merge remote-tracking branch 'upstream/main' into split-behavior

iff committed:
MovePane on single pane results in EmptyPane in opposite direction
- user intended to add an empty panel where the current item is and move the item to the other direction
- still needs testing and one check in the unit tests is missing

iff committed:
pane: cleanup tests

MrSubidubi committed:
Merge branch 'main' into split-behavior

MrSubidubi committed:
Formatting

MrSubidubi committed:
Format and tidy a bit

MrSubidubi reviewed (approved):
Nice. Thank you so much for bearing with me, absolutely awesome job on this. Good improvement, good code and a pleasure to discuss this with you.

Sorry for me being a bit unresponsive at the start and inbetween and thank you for being fine with it. Hope to see you again in another PR perhaps, other than that, happy holidays!

MrSubidubi committed:
And now imagine I could type...
